onnxruntime/core/optimizer/conv_add_fusion.cc

```diff
@@ -9,136 +9,108 @@ using namespace ONNX_NAMESPACE;
-    } else {
-      NodeArg* add_B_node_arg = graph.GetNodeArg(add_B_tensor_proto->name());
-      if (add_B_node_arg == nullptr) {
-        continue;
-      }
-
-      // Update shape of tensor proto
-      ONNX_NAMESPACE::TensorProto new_conv_B_tensor_proto(*add_B_tensor_proto);
-      int64_t dim = conv_W_tensor_proto->dims(0);
-      new_conv_B_tensor_proto.clear_dims();
-      new_conv_B_tensor_proto.add_dims(dim);
-
-      graph.RemoveInitializedTensor(add_B_tensor_proto->name());
-      graph.AddInitializedTensor(new_conv_B_tensor_proto);
-
-      // Update shape of NodeArg
-      TensorShapeProto shape;
-      shape.add_dim()->set_dim_value(dim);
-      add_B_node_arg->SetShape(shape);
-
-      conv_node.MutableInputDefs().push_back(add_B_node_arg);
-      conv_node.MutableInputArgsCount()[2] = 1;
+    // Calculate new value of initializers of conv node
+    conv_B->add(*add_B);
+
+    // Create new initializers of conv
+    ONNX_NAMESPACE::TensorProto new_conv_B_tensor_proto;
+    conv_B->ToProto(&new_conv_B_tensor_proto);
+
+    // Replace initializers of conv node
+    graph.RemoveInitializedTensor(conv_inputs[2]->Name());
+    graph.AddInitializedTensor(new_conv_B_tensor_proto);
+  } else {
+    NodeArg* add_B_node_arg = graph.GetNodeArg(add_B_tensor_proto->name());
+    if (add_B_node_arg == nullptr) {
+      return Status::OK();
     }
 
-    // Replace the input of the node following add node
-    const NodeArg* add_output_def = add_node.OutputDefs()[0];
-    NodeArg* conv_output_def = conv_node.MutableOutputDefs()[0];
-    for (auto it = add_node.OutputNodesBegin(); it != add_node.OutputNodesEnd(); ++it) {
-      auto output_node = graph.GetNode((*it).Index());
-      if (!output_node) {
-        return Status(ONNXRUNTIME, INVALID_ARGUMENT);
-      }
-      auto& input_defs = output_node->MutableInputDefs();
-      for (auto& def : input_defs) {
-        if (def == add_output_def) {
-          def = conv_output_def;
-        }
-      }
-    }
+    // Update shape of tensor proto
+    ONNX_NAMESPACE::TensorProto new_conv_B_tensor_proto(*add_B_tensor_proto);
+    int64_t dim = conv_W_tensor_proto->dims(0);
+    new_conv_B_tensor_proto.clear_dims();
+    new_conv_B_tensor_proto.add_dims(dim);
 
-    removed_nodes.push_back(add_node.Index());
-  }
+    graph.RemoveInitializedTensor(add_B_tensor_proto->name());
+    graph.AddInitializedTensor(new_conv_B_tensor_proto);
+
+    // Update shape of NodeArg
+    TensorShapeProto shape;
+    shape.add_dim()->set_dim_value(dim);
+    add_B_node_arg->SetShape(shape);
 
-  for (auto i : removed_nodes) {
-    graph.RemoveNode(i);
+    conv_node.MutableInputDefs().push_back(add_B_node_arg);
+    conv_node.MutableInputArgsCount()[2] = 1;
   }
 
-  if (!removed_nodes.empty()) {
+  // Remove Add node.
+  auto* add_node_to_remove = graph.GetNode(add_node.Index());
+  if (graph_utils::RemoveNode(graph, *add_node_to_remove)) {
     modified = true;
   }
```

Comment:
Should this set 'removed' to true when it removes the node?

Reply:
We use the removed to denote that the current node is removed, so that we stop the rule application for that node (if there are multiple rules for that node and the first one removes the node, then we should stop the rule application for that specific node).
However, here we are deleting the subsequent node (the Add), so it is fine to continue the rule application.
It is in my to-do list to substitute the two booleans with an enum, so that this becomes more clear.

Reply:
I think the loop in RuleBasedGraphTransformer needs updating. If we're removing the following node, we need to allow for GetNode to return a nullptr given the list of nodes being iterated will still include the deleted node. Currently it exits the loop. 

```
Status RuleBasedGraphTransformer::ApplyImpl(Graph& graph, bool& modified, int graph_level) const {
  GraphViewer graph_viewer(graph);
  auto& order = graph_viewer.GetNodesInTopologicalOrder();

  for (NodeIndex i : order) {
    auto* node = graph.GetNode(i);
    if (!node) {
      return Status(ONNXRUNTIME, INVALID_ARGUMENT);
    }
```

Reply:
I thought I have updated it already -- it shows up in the list of changes for me.

Reply:
In latest commit I also introduced the RewriteRuleEffect enum (let's do it, since we are at it now) -- let me know how it looks.